File: kompartment/python/kompartment/engine/partition.py
"""Cutting a model into parts that cannot see each other.

A port of the application's ``src/sim/partition.js``. A large assessment is
rarely one problem: its radionuclides fall into groups with no path between
them -- each decay chain held together by its own ingrowth and nothing joining
one chain to the next -- and solved as one system every state moves at the
step size the worst of them needs. This finds those groups; what to do with
them is :mod:`kompartment.engine.split`'s business.

**The graph is the Jacobian's pattern.** Two states belong together when one
can reach the other, and ``df/dy`` is exactly that relation: an entry at
(i, j) when the derivative of state i reads state j, through however many
expressions and rates. It is undirected here -- a path either way puts two
states together, since solving them apart needs them not to reach each other
in either direction.

**Where the pattern is not enough, the model is refused.** A snapshot and a
delay report the past, which no present state can move, so the Jacobian gives
them no column; an event moves a state its condition never mentions. A
dependency through any of them is invisible to the pattern, and a partition
that misses one does not run slowly -- it returns a different answer, quietly.
So a model with one is not split (:func:`why_not_split`). A min/max and a
running mean are not in that list: the Jacobian gives both a column through
their target, so the pattern already carries them.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

import numpy as np

__all__ = ['state_partition', 'why_not_split', 'partition_of']


def state_partition(pattern: Any) -> Dict[str, Any]:
    """The states of each part, from a sparsity pattern (``statePartition``).

    ``{'count', 'of', 'sizes', 'largest'}``: ``of[i]`` is the part state ``i``
    belongs to, numbered from 0 in the order the states first appear, so part
    0 holds state 0 -- a property of the model rather than of how the
    components were found.
    """
    n = int(getattr(pattern, 'n', 0) or 0) if pattern is not None else 0
    if n == 0:
        return {'count': 0, 'of': np.zeros(0, dtype=np.int64), 'sizes': np.zeros(0, dtype=np.int64), 'largest': 0}
    import scipy.sparse as sp
    from scipy.sparse.csgraph import connected_components
    col_ptr = np.asarray(pattern.col_ptr, dtype=np.int64)
    row_idx = np.asarray(pattern.row_idx, dtype=np.int64)
    graph = sp.csc_matrix((np.ones(row_idx.size, dtype=np.int8), row_idx, col_ptr), shape=(n, n))
    # The components of the undirected graph -- the application finds the same
    # ones by union-find -- then numbered in the order each one's first state
    # appears, which is the application's numbering and depends on nothing
    # but the model.
    count, labels = connected_components(graph, directed=True, connection='weak')
    first = np.full(count, n, dtype=np.int64)
    np.minimum.at(first, labels, np.arange(n, dtype=np.int64))
    renumber = np.empty(count, dtype=np.int64)
    renumber[np.argsort(first, kind='stable')] = np.arange(count, dtype=np.int64)
    of = renumber[labels].astype(np.int64)
    sizes = np.bincount(of, minlength=count).astype(np.int64)
    return {'count': count, 'of': of, 'sizes': sizes, 'largest': int(sizes.max()) if count else 0}


#: The blocks whose dependency the Jacobian pattern deliberately leaves out.
UNSEEN = {
    'delay': ('a delay reports the past, which no present state can move, so the Jacobian gives it no column '
              'and a dependency through it is invisible to this partition'),
    'snapshot': ('a snapshot reports the past, which no present state can move, so the Jacobian gives it no '
                 'column and a dependency through it is invisible to this partition'),
    'trigger': ('an event moves a state its condition never mentions, which is not a derivative and so is not '
                'in the Jacobian'),
}


def why_not_split(system: Any) -> Optional[str]:
    """Why this system may not be split, or None when it may (``whyNotSplit``).
    It names the first thing it cannot account for rather than guessing."""
    jac = getattr(system, 'jacobian', None) or {}
    if not jac.get('available'):
        return 'the model has no analytic Jacobian, so there is no sparsity pattern to read the parts off'
    pattern = jac.get('pattern')
    if pattern is None or not getattr(pattern, 'n', 0):
        return 'the model has no states to split'
    for rec in getattr(system, 'recorders', None) or []:
        why = UNSEEN.get(rec.kind)
        if why:
            return f"'{rec.name}' is a {rec.kind.replace('_', ' ')}: {why}"
    # A semi-analytical path's release is a convolution over what flowed into
    # it during the run, and the series of a split run are worked out
    # afterwards on the whole model's system, which never saw that history.
    layout = getattr(system, 'layout', None)
    for p in (layout.get('farfields') if layout is not None else None) or []:
        if p.farf.laplace:
            return (f"'{p.name}' is worked out semi-analytically: its release is a convolution over what flowed "
                    'into it during the run, which the whole model the series are worked out on afterwards does '
                    'not have')
    events = getattr(system, 'events', None)
    if events is not None and getattr(events, 'n', 0):
        many = 'an event' if events.n == 1 else f'{events.n} events'
        return (f'the model has {many}, and an event moves a state its condition never mentions, which is not a '
                'derivative and so is not in the Jacobian')
    return None


def partition_of(system: Any) -> Dict[str, Any]:
    """The partition, with the reason when there is not one (``partitionOf``):
    ``{'ok', 'refusal', 'count', 'of', 'sizes', 'largest'}``."""
    refusal = why_not_split(system)
    if refusal:
        return {'ok': False, 'refusal': refusal, 'count': 1, 'of': None, 'sizes': None, 'largest': 0}
    p = state_partition(system.jacobian['pattern'])
    return {'ok': True, 'refusal': None, **p}

File: kompartment/src/sim/partition.js
/**
 * Cutting a model into parts that cannot see each other.
 *
 * A large assessment is rarely one problem. model A carries
 * 22,842 states, and they fall into 32 groups with no path between them: the
 * radionuclides, with each decay chain held together by its own ingrowth and
 * nothing joining one chain to the next. Solved as one system, every state
 * moves at the step size the *worst* of them needs -- `ndf` picks it from an
 * infinity norm over all of them, so at any step exactly one
 * state chose it. On one 23,436-state model, 262 states ever set that step and half the
 * steps came from 26 of them. The other seventeen thousand are carried along.
 *
 * The idea is not new -- the desktop tools in this family call it batch mode,
 * and describe it as solving the equations for each contaminant separately,
 * with the exception that radionuclides of one decay chain stay together. That
 * description and this partition agree exactly on the corpus here: every state
 * of one such model carries a nuclide, no nuclide's states land in more than one part,
 * and the 32 parts hold 10, 6, 5, 3 and 3 nuclides with 27 alone -- which is a
 * list of decay chains. The graph is not told any of that; it finds it.
 *
 * **This file finds the partition. It does not decide what to do with one.**
 *
 * ---
 *
 * **Why the Jacobian pattern is the right graph, and where it is not enough.**
 *
 * Two states belong together when one can reach the other, and `df/dy` is
 * exactly that relation: the pattern has an entry at (i, j) when the
 * derivative of state i reads state j, through however many expressions and
 * transfer rates. It is already computed for the analytic Jacobian, it is
 * per-index rather than per-block, and the algebraic blocks between two states
 * are already collapsed into it.
 *
 * What it is not is a record of everything a model *does*. `jacobian.js` says
 * so where it handles the blocks that remember:
 *
 *     A snapshot and a delay report the past, which no present state can move,
 *     so neither gains a column.
 *
 * That is right for a Jacobian -- the derivative through a delay genuinely is
 * zero -- and wrong for this. A block reading a Delay of a compartment in
 * another part depends on it just as surely as a transfer would, and the
 * pattern shows nothing. An event is the same shape from the other end: its
 * action moves a state its condition never mentions. This is a known trap
 * rather than a theoretical one -- every implementation of batching that has
 * shipped has had to add a pass that merges the states an event, a min/max or
 * a snapshot reaches across.
 *
 * So a partition taken off the pattern alone is unsafe for a model with any of
 * those, and a wrong partition does not run slowly -- it returns a different
 * answer, quietly. `whyNotSplit` therefore refuses the model rather than
 * splitting it on an incomplete graph. What it refuses is narrow: of the
 * eleven assessment models of one series, not one carries a delay, a snapshot or an
 * event, so the pattern is their whole dependency graph and the partition is
 * provably complete. one assessment model carries 1 delay, 4 snapshots and
 * 14 discrete events, and is declined until the graph is widened to hold them.
 *
 * A min/max and a running mean are not in that list on purpose: `jacobian.js`
 * gives both a column through `rec.aux.target`, so the pattern already carries
 * them.
 */

/**
 * The states of each part, from the Jacobian's sparsity pattern.
 *
 * Undirected: a path either way puts two states together, because solving them
 * apart needs them not to reach each other in *either* direction. (The
 * directed version -- blocks that only feed downstream -- is a different and
 * weaker condition, and needs the interpolated coupling this does not.)
 *
 * @param {{n: number, colPtr: Int32Array|number[], rowIdx: Int32Array|number[]}} pattern
 * @returns {{count: number, of: Int32Array, sizes: Int32Array, largest: number}}
 *   `of[i]` is the part state i belongs to, numbered from 0 in the order the
 *   states first appear, so part 0 holds state 0.
 */
export function statePartition(pattern) {
	const n = pattern?.n ?? 0;
	const parent = new Int32Array(n);
	const rank = new Int32Array(n);
	for (let i = 0; i < n; i++) parent[i] = i;

	// Path halving: every other node on the way up is pointed at its
	// grandparent, which flattens the tree without a second pass.
	const find = (x) => {
		while (parent[x] !== x) {
			parent[x] = parent[parent[x]];
			x = parent[x];
		}
		return x;
	};
	const union = (a, b) => {
		a = find(a);
		b = find(b);
		if (a === b) return;
		if (rank[a] < rank[b]) { const t = a; a = b; b = t; }
		parent[b] = a;
		if (rank[a] === rank[b]) rank[a]++;
	};

	const { colPtr, rowIdx } = pattern ?? {};
	for (let j = 0; j < n; j++) {
		for (let k = colPtr[j]; k < colPtr[j + 1]; k++) union(j, rowIdx[k]);
	}

	// Number the parts in the order their first state appears, so the numbering
	// is a property of the model rather than of the union-find's internals.
	const label = new Int32Array(n).fill(-1);
	const of = new Int32Array(n);
	let count = 0;
	for (let i = 0; i < n; i++) {
		const r = find(i);
		if (label[r] < 0) label[r] = count++;
		of[i] = label[r];
	}

	const sizes = new Int32Array(count);
	for (let i = 0; i < n; i++) sizes[of[i]]++;
	let largest = 0;
	for (let c = 0; c < count; c++) if (sizes[c] > largest) largest = sizes[c];

	return { count, of, sizes, largest };
}

/** The blocks whose dependency the Jacobian pattern deliberately leaves out. */
const UNSEEN = {
	delay: 'a delay reports the past, which no present state can move, so the '
		+ 'Jacobian gives it no column and a dependency through it is invisible '
		+ 'to this partition',
	snapshot: 'a snapshot reports the past, which no present state can move, so '
		+ 'the Jacobian gives it no column and a dependency through it is '
		+ 'invisible to this partition',
	trigger: 'an event moves a state its condition never mentions, which '
		+ 'is not a derivative and so is not in the Jacobian',
};

/**
 * Why this system may not be split, or null when it may.
 *
 * Conservative by design: it names the first thing it cannot account for
 * rather than guessing. A model that is refused is solved whole, which is what
 * every model here did until now.
 *
 * @param {object} system  from `buildSystem`
 * @returns {string|null}
 */
export function whyNotSplit(system) {
	if (!system?.jacobian?.available) {
		return 'the model has no analytic Jacobian, so there is no sparsity '
			+ 'pattern to read the parts off';
	}
	if (!system.jacobian.pattern?.n) return 'the model has no states to split';

	for (const rec of system.recorders ?? []) {
		const why = UNSEEN[rec.kind];
		if (why) return `'${rec.name}' is a ${rec.kind.replace(/_/g, ' ')}: ${why}`;
	}
	// A semi-analytical path's release is a convolution over what flowed into
	// it during the run, and the series of a split run are worked out
	// afterwards on the whole model's system, which never saw that history.
	const laplace = (system.layout?.farfields ?? []).find((p) => p.farf?.laplace);
	if (laplace) {
		return `'${laplace.name}' is worked out semi-analytically: its release is a convolution `
			+ 'over what flowed into it during the run, which the whole model the series are '
			+ 'worked out on afterwards does not have';
	}
	// Events arrive as a compiled bundle rather than as recorders when the
	// model declares them that way, so both routes are checked.
	if (system.events?.n) {
		return `the model has ${system.events.n === 1 ? 'an event' : `${system.events.n} events`}`
			+ ', and an event moves a state its condition never mentions, which is '
			+ 'not a derivative and so is not in the Jacobian';
	}
	return null;
}

/**
 * The partition, with the reason when there is not one.
 *
 * @param {object} system  from `buildSystem`
 * @returns {{ok: boolean, refusal: string|null, count: number,
 *   of: Int32Array|null, sizes: Int32Array|null, largest: number}}
 */
export function partitionOf(system) {
	const refusal = whyNotSplit(system);
	if (refusal) {
		return { ok: false, refusal, count: 1, of: null, sizes: null, largest: 0 };
	}
	const p = statePartition(system.jacobian.pattern);
	return { ok: true, refusal: null, ...p };
}
